addons/libs/images.lua

```diff
@@ -244,6 +246,15 @@ function images.pos_y(t, y)
     t:pos(meta[t].settings.pos.x, y)
 end
 
+function images.last_pos(t, x, y)
+    if x == nil then
+        return last_pos.x, last_pos.y
+    end
+
+    last_pos.x = x
+    last_pos.y = y
+end
```

Comment:
As discussed, this should be removed. The local `last_pos` variable should be used instead, similar to the `dragged` variable.

Reply:
Please advise how calculating the drag delta can be accomplished from the addon's scope without this. My delta calculations fail without this, as my addon sees last_pos as a nil value and don't appear to be able to access last_pos at all. This bridges the gap. Is there another way?

Reply:
Addons should not calculate deltas. That is the lib's job. I'm still not convinced about the validity of this solution for your use case, but even in this case, the deltas could be sent as event parameters. I feel like this could be handled better in general, though. I'll have to think about this.

Reply:
AFAIK, when people are watching drag events, they are using it to keep objects in sync. In such a case, they are using a delta in one form or another.

**A. Top-Left Delta:** _(More complicated)_
1. Find the current image's x/y distance from a bounding rectangle's top-left corner. (one that covers all the shapes)
2. Subtract x/y distance from the last-known position of the bounding rectangle to get the movement delta.

**B. Mouse-Movement Delta** _(Simpler)_
1. Get the last known mouse x/y position
2. Subtract it from the current mouse x/y position to get the delta.

If drag events are going to be doing those calculations either way - _meaning the same overhead_ -  and getting the same result - _the same exact x/y delta, just calculated two different ways_ - it may be nice to provide the simpler method. Same overhead, but seems more user friendly, a win/win for service delivery and customer satisfaction.

Passing it in as an event arg sounds convenient. Would that muck with anything else, though?

Reply:
> AFAIK, when people are watching drag events, they are using it to keep objects in sync.

You're literally the only one doing this, because this feature doesn't even exist yet :P

> Passing it in as an event arg sounds convenient for the addon authors. Would that muck with anything else, though?

No, but I still don't like it, this could be handled entirely in addon code. You'd have to keep track of the objects' original positions, then compare against the current, and adjust accordingly. Even with multiple directly draggable objects it wouldn't be too hard to do.

The reason this bothers me so much is because this feels out of scope for this lib, it's what a widget lib would be for. This lib only concerns individual objects, nothing outside of it, so your use-case of synchronizing with other elements feels like a weak justification.

As for your scenarios, only B makes sense. A is not more reliable. We only report discrete pixel values and they're fully accurate (to the pixel level).

Reply:
> > AFAIK, when people are watching drag events, they are using it to keep objects in sync.
> 
> You're literally the only one doing this, because this feature doesn't even exist yet :P

Why do you say that I'm the only one who syncs objects during drag? [As can be seen here,](https://discord.com/channels/338590234235371531/501099842097905675/994989166377050253) Akirane also does it, and there are likely others that do it as well that we're not aware of. It's not factual to say that I'm the only one, unless you're going on the technicality of _"using image:register_event('drag', fn)"_ vs _"windower.register_event('mouse', fn)"_, which would be missing the point if so.

As far as I can tell, if someone's watching drag then they'll be interested in the deltas - _that's what Akirane, probably unknown others, I are using the drag event for_ - as I don't yet see what else is important for an addon author during a drag.

> > Passing it in as an event arg sounds convenient for the addon authors. Would that muck with anything else, though?
> 
> No[...]

Great! Well, if the drag delta is the only important thing during a drag event, and we strip that delta from the event, we may be gutting the drag event from its usefulness.

> The reason this bothers me so much is because this feels out of scope for this lib, it's what a widget lib would be for. This lib only concerns individual objects, nothing outside of it, so your use-case of synchronizing with other elements feels like a weak justification.

Would a widget lib let the author bind (or whatever technical term) a bunch of images/texts into a cluster and then the widget lib would move them in tandem? That seems like a nice lib to have! I would agree that, if something like that is offered, that would seem like a better place for this. But, until then, what? The drag event could meet people's needs, if we build it to provide them the delta.

> We only report discrete pixel values and they're fully accurate (to the pixel level).

Great to hear!

Reply:
Akirane does the same thing in XIVHotBar. However, this does seem like it would better belong in a (not-yet-available) widgets lib, like you said. I guess for the meantime such authors should continue doing it themselves, right?